socorro/cron/jobs/monitoring.py

```diff
@@ -0,0 +1,171 @@
+
+
+REPO_ROOT = dirname(dirname(dirname(dirname(__file__))))
+
+
+VulnerabilityBase = namedtuple('Vulnerability', (
+    'type',
+    'dependency',
+    'installed_version',
+    'affected_versions',
+    'description',
+))
+
+
+class Vulnerability(VulnerabilityBase):
+    @property
+    def key(self):
+        return '[%s] %s' % (self.type, self.dependency)
+
+    @property
+    def summary(self):
+        return 'Installed: %s; Affected: %s; %s' % (
+            self.installed_version,
+            self.affected_versions,
+            self.description,
+        )
+
+
+class DependencySecurityCheckFailed(Exception):
+    """Thrown when a security check cannot complete, such as network
+    issues.
+
+    """
+
+
+class DependencySecurityCheckCronApp(BaseCronApp):
+    """Configuration values used by this app:
+
+    crontabber.class-DependencySecurityCheckCronApp.node_modules
+        Path to the node_modules directory where the webapp's npm
+        dependencies have been installed.
+    secrets.sentry.dsn
+        If specified, vulnerabilities will be reported to Sentry instead
+        of logged to the console.
+
+    """
+    app_name = 'dependency-security-check'
+    app_description = (
+        'Runs third-party tools that check for known security vulnerabilites in Socorro\'s '
+        'dependencies.'
+    )
+    app_version = '0.1'
+
+    required_config = Namespace()
+    required_config.add_option(
+        'node_modules',
+        doc=(
+            'Path to node_modules directory where the webapp\'s npm dependencies have been '
+            'installed.'
+        ),
+    )
+
+    def run(self):
+        vulnerabilities = self.get_python_vulnerabilities() + self.get_javascript_vulnerabilities()
+        if vulnerabilities:
+            try:
+                dsn = self.config.sentry.dsn
+            except KeyError:
+                dsn = None
+
+            if dsn:
+                self.alert_sentry(dsn, vulnerabilities)
+            else:
+                self.alert_log(vulnerabilities)
+
+    def alert_sentry(self, dsn, vulnerabilities):
+        client = raven_client.get_client(dsn)
+        client.context.activate()
+        client.context.merge({
+            'extra': {
+                'data': {vuln.key: vuln.summary for vuln in vulnerabilities},
+            },
+        })
+        client.captureMessage('Dependency security check failed')
+
+    def alert_log(self, vulnerabilities):
+        for vuln in vulnerabilities:
+            self.config.logger.error('%s: %s' % (vuln.key, vuln.summary))
+
+    def get_python_vulnerabilities(self):
+        """Check Python dependencies via Pyup's safety command.
+
+        :returns list(Vulnerability):
+        :raises DependencySecurityCheckFailed:
+        """
+        # Safety checks what's installed in the current virtualenv, so no need
+        # for any paths.
+        process = Popen(['safety', 'check', '--json'], stdin=PIPE, stdout=PIPE, stderr=PIPE)
```

Comment:
We should use an absolute path for `safety` so we know exactly which one is being executed.

Bandit says this:

```
--------------------------------------------------
>> Issue: [B607:start_process_with_partial_path] Starting a process with a partial executable path
   Severity: Low   Confidence: High
   Location: socorro/cron/jobs/monitoring.py:111
110	        # for any paths.
111	        process = Popen(['safety', 'check', '--json'], stdin=PIPE, stdout=PIPE, stderr=PIPE)
112	        output, error_output = process.communicate()
```

Reply:
Should we just switch to having `nsp_path` and `safety_path` options?

Reply:
That's a good way to solve it! +1